message_endpoint.go

```diff
@@ -0,0 +1,314 @@
+	messageHandler MessageHandler
+	awareness      *Awareness
+	resHandler     *responseHandler
+	quit           chan struct{}
+}
+
+func NewMessageEndpoint(config MessageEndpointConfig, transport UDPTransport, messageHandler MessageHandler, awareness *Awareness) (*MessageEndpoint, error) {
+	if config.CallbackCollectInterval == time.Duration(0) {
+		return nil, ErrCallbackCollectIntervalNotSpecified
+	}
+
+	return &MessageEndpoint{
+		config:         config,
+		transport:      transport,
+		messageHandler: messageHandler,
+		awareness:      awareness,
+		resHandler:     newResponseHandler(config.CallbackCollectInterval),
+		quit:           make(chan struct{}),
+	}, nil
+}
+
+// Listen is a log running goroutine that pulls packet from the
+// transport and pass it for processing
+func (m *MessageEndpoint) Listen() {
+	for {
+		select {
+		case packet := <-m.transport.PacketCh():
+			// validate packet then convert it to message
+			msg, err := m.processPacket(*packet)
+			if err != nil {
+				iLogger.Panic(nil, err.Error())
+			}
+
+			// before message that come from other handle by MessageHandler
+			// check whether this message is sent-back message from other member
+			// this is determined by message's Seq property which work as message id
+
+			if m.resHandler.hasCallback(msg.Seq) {
+				go m.resHandler.handle(msg)
+			} else {
+				go m.handleMessage(msg)
+			}
+
+		case <-m.quit:
+			return
+		}
+	}
+}
+
+// ProcessPacket process given packet, this procedure may include
+// decrypting data and converting it to message
+func (m *MessageEndpoint) processPacket(packet Packet) (pb.Message, error) {
+	msg := &pb.Message{}
+	if m.config.EncryptionEnabled {
+		// TODO: decrypt packet
+	}
+
+	if err := proto.Unmarshal(packet.Buf, msg); err != nil {
+		return pb.Message{}, err
+	}
+
+	return *msg, nil
+}
+
+func (m *MessageEndpoint) Shutdown() {
+	// close transport first
+	m.transport.Shutdown()
+
+	// then close message endpoint
+	m.quit <- struct{}{}
+}
+
+func validateMessage(msg pb.Message) bool {
+	if msg.Seq == "" {
+		iLogger.Info(nil, "message seq value empty")
+		return false
+	}
+
+	if msg.Payload == nil {
+		iLogger.Info(nil, "message payload value empty")
+		return false
+	}
+
+	return true
+}
+
+// with given message handleMessage determine which logic should be executed
+// based on the message type. Additionally handleMessage can call MemberDelegater
+// to update member status and encrypt messages
+func (m *MessageEndpoint) handleMessage(msg pb.Message) error {
+	// validate message
+	if !validateMessage(msg) {
+		return ErrInvalidMessage
+	}
+
+	// call delegate func to update members states
+	m.messageHandler.handle(msg)
+	return nil
+
+}
+
+// determineSendTimeout if @timeout is given, then use this value as timeout value
+// otherwise calculate timeout value based on the awareness
+func (m *MessageEndpoint) determineSendTimeout(timeout time.Duration) time.Duration {
+	if timeout != DefaultSendTimeout {
```

Comment:
If i want to set the timer to the config value, do i have to pass 0 as an argument? And if i set timeout through argument, is it right that not using awareness?